language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs

```diff
@@ -47,6 +54,108 @@ updated state that the first 'eval' finished with. This may cause all kinds of w
 for example, an error message saying that there is a free variable and evaluation cannot proceed.
 -}
 
+-- | A wrapper around 'Term' parameterized by a phantom type variable representing the Haskell
+-- counterpart of the type of that term.
+newtype TypedTerm a = TypedTerm
+    { unTypedTerm :: Term TyName Name ()
+    }
+
+{- Note [Dynamic builtins as constructors]
+We can use dynamic builtins as improvised constructors of data types.
+Consider the 'Sealed' example, we have:
+
+1. the hardcoded @sealed@ PLC type of kind @* -> *@
+2. the @seal@ dynamic builtin of type @all a. a -> sealed a@
+3. the @unseal@ dynamic builtin of type @all a. sealed a -> a@
+
+The Haskell meanings of these things are:
+
+1. 'Sealed' (data type)
+2. 'Sealed' (constructor)
+3. 'unSealed' (field selector)
+
+I.e.
+
+    seal {Integer} 1
+
+is a value of type @sealed integer@, i.e. @seal@ plays the role of a constructor here,
+even though it's really a dynamic builtin.
+
+We immediately run into problems, though. As a dynamic builtin, @seal@ has a meaning and thus
+computes. So @seal@ becomes @Sealed@ on the Haskell side, but we also need to go in the other
+direction, meaning that @Sealed@ becomes @seal@ back. I.e. the
+
+    seal {Integer} 1
+
+application evaluates to itself, which means that the previous version of the CEK machine was
+looping while trying to evaluate that application over and over again. The hack that we added
+to handle this is trivial: if an application evaluates to itself, do not attempt to evaluate it
+again.
```

Comment:
Hah, I had exactly this problem, but I didn't consider adding a hack :see_no_evil: